src/textformatter.cpp

```diff
@@ -61,23 +61,36 @@ std::vector<std::string> wrap_line(
 	std::string curline = "";
 
 	for (auto word : words) {
-		size_t word_length = utils::strwidth_stfl(word);
-		size_t curline_length = utils::strwidth_stfl(curline);
-
-		// For words wider than the available width we have no other choice but
-		// to force splits at width limit
-		while (word_length > width) {
-			size_t space_left = width - curline_length;
-			curline.append(word.substr(0, space_left));
-			word.erase(0, space_left);
+		size_t word_width = utils::strwidth_stfl(word);
+		size_t curline_width = utils::strwidth_stfl(curline);
+
+		// for languages (e.g., CJK) don't use a space as a word boundary
+		while (word_width > width) {
+			size_t space_left = width - curline_width;
+			std::string part = utils::substr_with_width(word, space_left);
+			curline.append(part);
+			word.erase(0, part.length());
+			if (word.empty()) {
+				word_width = 0;
+				break;
+			}
 			result.push_back(curline);
 			curline = prefix;
 
-			word_length = utils::strwidth_stfl(word);
-			curline_length = utils::strwidth_stfl(curline);
+			word_width = utils::strwidth_stfl(word);
+			curline_width = utils::strwidth_stfl(curline);
+			space_left = width - curline_width;
+
+			// if the new line have no enough space to put single wide char,
+			// discard the current word as a workaround
+			if (space_left < 2) {
```

Comment:
1. How can we be sure that wide character occupies two places?
2. Isn't discarding the current word a bad thing? Are you sure we never lose any text this way?
3. You mention it's a wordaround, but never tell which problem it solves. Can you add a test demonstrating a situation where without this check, the test will fail?